### ISSUE TITLE
Code pattern for constructing urls

### ISSUE THREAD

**Issue by bvw** [qcode-software/qcode-tcl#300]
Acceptance Criteria: Agree a standard code pattern for constructing restful urls optionally with GET string form variables.

**Comment by bvw** [comment]
I'm content with the current implementation of qc::url

1. Backward compatible with existing codebase
2. Supports colon variable substitution into url path
3. Supports GET string form variables with unique names.(feature not bug)

**Comment by PeterChaplin** [comment]
My concern is that `qc::url` is currently more verbose and less readable than `qc::url_maker`, when creating RESTful urls from local vars (which, moving forwards, is likely to be the main use-case for url creation).

There's also a minor bug in `qc::url`, eg
```tcl
register GET /test {next_url} {
    return [qc::url $next_url foo bar]
}
```
with the requested url
```
/test?next_url=/somewhere/:foo
```
Not a major problem, but I think it might be worth looking at a way to pass a base url into `qc::url` without it being subject to colon substitution, but also without it being subject to url-escaping. For example, if the url passed in with a colon, we could turn off url escaping, allowing us to write
```tcl
qc::url :next_url foo bar
```
and generate
```
/somewhere/:foo?foo=bar
```

**Comment by PeterChaplin** [comment]
We're already switching from proc to register, and introducing RESTful urls, so whatever we decide is going to establish a new pattern. The existing codebase doesn't have a pattern for creating RESTful urls, because we haven't been using them, and I'm not sure that a developer who is familiar with existing `url` usage is going to find colon substitution easier to learn than `url_maker`.

**Comment by nicky-johnstone** [comment]
What I prefer about `qc::url` over `qc::url_maker` is that the pattern passed in is more closely representative of the final URL.

```
url /foo/bar/:baz ~ baz
```

 It also matches the URL patterns used for `qc::register`:

```
qc::register GET /foo/bar/:baz {baz} {

}
```

Also, if the URL pattern in the above handler is changed then I'd think it more intuitive to search for other occurrences of the same pattern whereas with `qc::url_maker` you'd have to search for `/ foo bar :baz`. This point is likely minor or irrelevant though because once the standard is established then we'd know how to search for URL patterns if the handler were to change.

**Comment by PeterChaplin** [comment]
Agreed. If we could get colon var substitution from local vars working without introducing a security hole, I'd consider that more-or-less ideal.
